compiler/parser/Parser.y

```diff
@@ -3400,30 +3403,59 @@ qual  :: { LStmt GhcPs (LHsExpr GhcPs) }
+fbinds1 :: { ([AddAnn],([Fbind (HsExpr GhcPs)], Bool)) }
         : fbind ',' fbinds1
-                {% addAnnotation (gl $1) AnnComma (gl $2) >>
+                {% let gl' = \x -> case x of {Left (L l _) -> l; Right (L l _) -> l} in
+                   addAnnotation (gl' $1) AnnComma (gl $2) >>
                    return (case $3 of (ma,(flds, dd)) -> (ma,($1 : flds, dd))) }
         | fbind ';' fbinds1
-                {% addAnnotation (gl $1) AnnComma (gl $2) >>
+                {% let gl' = \x -> case x of {Left (L l _) -> l; Right (L l _) -> l} in
+                   addAnnotation (gl' $1) AnnComma (gl $2) >>
                    return (case $3 of (ma,(flds, dd)) -> (ma,($1 : flds, dd))) }
         | fbind                         { ([],([$1], False)) }
         | '..'                          { ([mj AnnDotdot $1],([],   True)) }
 
-fbind   :: { LHsRecField GhcPs (LHsExpr GhcPs) }
-        : qvar '=' texp {% ams  (sLL $1 $> $ HsRecField (sL1 $1 $ mkFieldOcc $1) $3 False)
-                                [mj AnnEqual $2] }
+fbind   :: { Fbind (HsExpr GhcPs) }
+        : qvar '=' texp {% fmap Left $ ams  (sLL $1 $> $ HsRecField (sL1 $1 $ mkFieldOcc $1) $3 False)
+                                        [mj AnnEqual $2] }
                         -- RHS is a 'texp', allowing view patterns (Trac #6038)
                         -- and, incidentally, sections.  Eg
                         -- f (R { x = show -> s }) = ...
 
-        | qvar          { sLL $1 $> $ HsRecField (sL1 $1 $ mkFieldOcc $1) placeHolderPunRhs True }
+        | qvar          { Left $ sLL $1 $> $ HsRecField (sL1 $1 $ mkFieldOcc $1) placeHolderPunRhs True }
                         -- In the punning case, use a place-holder
                         -- The renamer fills in the final value
+        | field '.' fieldToUpdate '=' texp
+                        { do
+                            let top = $1
+                                fields = top : reverse $3
+                                final = last fields
+                                l = comb2 top final
+                                isPun = False
+                            Right $ mkRdrProjUpdate (comb2 $1 $5) (L l fields) $5 isPun
+                        }
+        | field '.' fieldToUpdate
+                        { do
+                          let top = $1
+                              fields = top : reverse $3
+                              final = last fields
+                              l = comb2 top final
+                              isPun = True
+                              var = mkUnqualVar $ mkVarOcc $ unpackFS $ unLoc final
+                          Right $ mkRdrProjUpdate l (L l fields) var isPun
+                        }
+
+fieldToUpdate :: { [Located FastString] }
+fieldToUpdate
+        : fieldToUpdate '.' field { $3 : $1 }
+        | field { [$1] }
+
+field :: { Located FastString }
+      : VARID { sL1 $1 $! getVARID $1 }
```

Comment:
do you know why this is a `VARID` and not a `qvar`? I think I don’t understand why the non-nested case has a different definition of the individual pieces.

Reply:
No I'm not sure why `VARID` was chosen over `qvar` here.